File: typescript/packages/core/src/event-bus.ts
import type { EventBus, StreamEvent } from './types/contracts.js';

export class InMemoryEventBus implements EventBus {
  private listeners = new Set<(event: StreamEvent) => void>();

  subscribe(listener: (event: StreamEvent) => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  publish(event: StreamEvent): void {
    for (const listener of this.listeners) {
      listener(event);
    }
  }
}

File: python/src/agent_framework/core/event_bus.py
from __future__ import annotations

from agent_framework.core.contracts import StreamEvent


class InMemoryEventBus:
    def __init__(self) -> None:
        self._listeners: set = set()

    def subscribe(self, listener):
        self._listeners.add(listener)
        return lambda: self._listeners.discard(listener)

    def publish(self, event: StreamEvent) -> None:
        for listener in list(self._listeners):
            listener(event)
